File: src/progression_rules.py
"""
Deterministic progression directives derived from prior-week logs.
"""

import re

from src.workout_db import normalize_exercise_name
from src.exercise_normalizer import get_normalizer


RPE_VALUE_RE = re.compile(r"\brpe\s*[:=]?\s*(\d+(?:\.\d+)?)\b", re.IGNORECASE)

HOLD_LOCK_PATTERNS = [
    re.compile(pattern, re.IGNORECASE)
    for pattern in [
        r"\bkeep\b",
        r"\bstay here\b",
        r"\bhold\b",
        r"\bsame weight\b",
        r"\bdon't increase\b",
        r"\bdo not increase\b",
        r"\bcan't increase\b",
    ]
]

STRUGGLE_PATTERNS = [
    re.compile(pattern, re.IGNORECASE)
    for pattern in [
        r"\bhard\b",
        r"\bheavy\b",
        r"\btough\b",
        r"\bstruggl",
        r"\bchalleng",
        r"\bfailed\b",
        r"\bform (?:broke|breakdown|wasn't perfect)\b",
    ]
]

EXCEEDED_PATTERNS = [
    re.compile(pattern, re.IGNORECASE)
    for pattern in [
        r"\beasy\b",
        r"\btoo light\b",
        r"\bcould do more\b",
        r"\breps? left\b",
        r"\bgo up\b",
        r"\bincrease\b",
    ]
]


def _parse_float(value):
    try:
        return float(str(value).strip())
    except (TypeError, ValueError):
        return None


def _parse_int(value):
    try:
        return int(str(value).strip())
    except (TypeError, ValueError):
        return None


def _parse_rpe(text):
    match = RPE_VALUE_RE.search(text or "")
    if not match:
        return None

    value = float(match.group(1))
    if 1.0 <= value <= 10.0:
        return value
    return None


def _matches_any(patterns, text):
    for pattern in patterns:
        if pattern.search(text or ""):
            return True
    return False


def _classify_signal(log_text, rpe_value):
    text = log_text or ""

    if _matches_any(HOLD_LOCK_PATTERNS, text):
        return "hold_lock", "explicit_keep_instruction"

    if _matches_any(STRUGGLE_PATTERNS, text):
        return "hold_lock", "struggle_signal"

    if rpe_value is not None and rpe_value >= 9.0:
        return "hold_lock", "high_rpe"

    if _matches_any(EXCEEDED_PATTERNS, text):
        return "progress", "exceeded_signal"

    if rpe_value is not None and rpe_value <= 7.0:
        return "progress", "low_rpe"

    return "neutral", "no_strong_signal"


def _normalize_day_name(day_name):
    return (day_name or "").strip().lower()


def _safe_str(value):
    return str(value).strip() if value is not None else ""


def build_progression_directives(prior_supplemental):
    """
    Build deterministic directives from prior-week supplemental logs.

    Returns:
        List[dict] with keys:
            day_name, exercise_name, normalized_exercise,
            signal, reason, hold_lock, target_reps, target_load,
            parsed_rpe, source_log
    """
    directives = []
    if not prior_supplemental:
        return directives

    for day_name in ["Tuesday", "Thursday", "Saturday"]:
        for exercise in prior_supplemental.get(day_name, []):
            exercise_name = _safe_str(exercise.get("exercise"))
            if not exercise_name:
                continue

            log_text = _safe_str(exercise.get("log"))
            if not log_text:
                continue

            parsed_rpe = _parse_rpe(log_text)
            signal, reason = _classify_signal(log_text, parsed_rpe)
            target_reps = _parse_int(exercise.get("reps"))
            target_load = _parse_float(exercise.get("load"))
            hold_lock = signal == "hold_lock"

            directives.append(
                {
                    "day_name": _normalize_day_name(day_name),
                    "exercise_name": exercise_name,
                    "normalized_exercise": normalize_exercise_name(exercise_name),
                    "signal": signal,
                    "reason": reason,
                    "hold_lock": hold_lock,
                    "target_reps": target_reps,
                    "target_load": target_load,
                    "parsed_rpe": parsed_rpe,
                    "source_log": log_text,
                }
            )

    return directives


def format_directives_for_prompt(directives, max_lines=18):
    """Render compact directive block for prompt context."""
    if not directives:
        return ""

    lines = ["PROGRESSION DIRECTIVES (DETERMINISTIC FROM PRIOR LOGS):"]
    hold_lines = []
    progress_lines = []

    for directive in directives:
        name = directive["exercise_name"]
        day = directive["day_name"].capitalize()
        log_preview = directive["source_log"][:90]

        if directive["hold_lock"]:
            reps = directive["target_reps"]
            load = directive["target_load"]
            hold_lines.append(
                f"- LOCK {day} | {name} -> keep {reps} reps @ {format_load(load)} kg | log: {log_preview}"
            )
        elif directive["signal"] == "progress":
            progress_lines.append(
                f"- PROGRESS {day} | {name} -> progression allowed (single-variable change) | log: {log_preview}"
            )

    selected = hold_lines + progress_lines
    lines.extend(selected[:max_lines])
    return "\n".join(lines)


def format_load(value):
    """Format load values while preserving meaningful decimal precision."""
    if value is None:
        return ""

    if abs(value - round(value)) < 1e-9:
        return str(int(round(value)))

    # Keep up to 3 decimals to support cable/machine increments.
    return f"{value:.3f}".rstrip("0").rstrip(".")


def _normalize_for_match(name):
    return get_normalizer().canonical_key(name)


def _find_best_directive(day_name, exercise_name, directive_map):
    normalizer = get_normalizer()
    target_day = _normalize_day_name(day_name)

    # Exact canonical key match
    key = (target_day, normalizer.canonical_key(exercise_name))
    if key in directive_map:
        return directive_map[key]

    # Fuzzy match via normalizer
    for (day_key, exercise_key), directive in directive_map.items():
        if day_key != target_day:
            continue
        if normalizer.are_same_exercise(exercise_name, directive.get("exercise_name", "")):
            return directive
    return None


def apply_locked_directives_to_plan(plan_text, directives):
    """
    Apply deterministic hold-lock directives to generated plan text.

    Returns:
        Tuple[str, int] => (updated_plan_text, num_applied)
    """
    if not plan_text or not directives:
        return plan_text, 0

    locked_directives = [d for d in directives if d.get("hold_lock")]
    if not locked_directives:
        return plan_text, 0

    normalizer = get_normalizer()
    directive_map = {}
    for directive in locked_directives:
        key = (
            _normalize_day_name(directive.get("day_name")),
            normalizer.canonical_key(directive.get("exercise_name")),
        )
        directive_map[key] = directive

    day_re = re.compile(r"^\s*##\s+([A-Z]+DAY)\b", re.IGNORECASE)
    exercise_re = re.compile(r"^\s*###\s+[A-Z]\d+\.\s*(.+)$", re.IGNORECASE)
    prescription_re = re.compile(
        r"^(\s*-\s*)(\d+)\s*x\s*([\d:]+)\s*@\s*([\d]+(?:\.\d+)?)\s*kg(\b.*)$",
        re.IGNORECASE,
    )

    current_day = None
    current_exercise = None
    current_directive = None
    applied = 0
    lines = plan_text.split("\n")

    for idx, line in enumerate(lines):
        day_match = day_re.match(line)
        if day_match:
            current_day = day_match.group(1).capitalize()
            current_exercise = None
            current_directive = None
            continue

        exercise_match = exercise_re.match(line)
        if exercise_match:
            current_exercise = exercise_match.group(1).strip()
            current_directive = _find_best_directive(current_day, current_exercise, directive_map)
            continue

        if not current_directive:
            continue

        prescription_match = prescription_re.match(line)
        if not prescription_match:
            continue

        prefix, sets, reps, _load, suffix = prescription_match.groups()
        target_reps = current_directive.get("target_reps")
        target_load = current_directive.get("target_load")
        if target_reps is None or target_load is None:
            continue

        new_line = f"{prefix}{sets} x {target_reps} @ {format_load(target_load)} kg{suffix}"
        if new_line != line:
            lines[idx] = new_line
            applied += 1

        # Only apply once per exercise prescription line.
        current_directive = None

    return "\n".join(lines), applied

File: native/SamsWorkoutNative/Sources/WorkoutCore/ProgressionRules.swift
import Foundation

public struct PriorSupplementalExercise: Equatable, Sendable {
    public let exercise: String
    public let reps: String
    public let load: String
    public let log: String

    public init(exercise: String, reps: String, load: String, log: String) {
        self.exercise = exercise
        self.reps = reps
        self.load = load
        self.log = log
    }
}

public struct ProgressionRuleDirective: Equatable, Sendable {
    public let dayName: String
    public let exerciseName: String
    public let normalizedExercise: String
    public let signal: String
    public let reason: String
    public let holdLock: Bool
    public let targetReps: Int?
    public let targetLoad: Double?
    public let parsedRPE: Double?
    public let sourceLog: String

    public init(
        dayName: String,
        exerciseName: String,
        normalizedExercise: String,
        signal: String,
        reason: String,
        holdLock: Bool,
        targetReps: Int?,
        targetLoad: Double?,
        parsedRPE: Double?,
        sourceLog: String
    ) {
        self.dayName = dayName
        self.exerciseName = exerciseName
        self.normalizedExercise = normalizedExercise
        self.signal = signal
        self.reason = reason
        self.holdLock = holdLock
        self.targetReps = targetReps
        self.targetLoad = targetLoad
        self.parsedRPE = parsedRPE
        self.sourceLog = sourceLog
    }

    public func asPlanDirective() -> ProgressionDirective {
        ProgressionDirective(
            dayName: dayName,
            exerciseName: exerciseName,
            holdLock: holdLock,
            targetReps: targetReps.map(String.init),
            targetLoad: targetLoad
        )
    }
}

private let rpeValueRegex = makeRegex("\\brpe\\s*[:=]?\\s*(\\d+(?:\\.\\d+)?)\\b", options: [.caseInsensitive])

private let holdLockPatterns: [NSRegularExpression] = [
    "\\bkeep\\b",
    "\\bstay here\\b",
    "\\bhold\\b",
    "\\bsame weight\\b",
    "\\bdon't increase\\b",
    "\\bdo not increase\\b",
    "\\bcan't increase\\b",
].map { makeRegex($0, options: [.caseInsensitive]) }

private let strugglePatterns: [NSRegularExpression] = [
    "\\bhard\\b",
    "\\bheavy\\b",
    "\\btough\\b",
    "\\bstruggl",
    "\\bchalleng",
    "\\bfailed\\b",
    "\\bform (?:broke|breakdown|wasn't perfect)\\b",
].map { makeRegex($0, options: [.caseInsensitive]) }

private let exceededPatterns: [NSRegularExpression] = [
    "\\beasy\\b",
    "\\btoo light\\b",
    "\\bcould do more\\b",
    "\\breps? left\\b",
    "\\bgo up\\b",
    "\\bincrease\\b",
].map { makeRegex($0, options: [.caseInsensitive]) }

private let progressionDayOrder = ["Tuesday", "Thursday", "Saturday"]

private func parseFloat(_ value: String) -> Double? {
    Double(value.trimmingCharacters(in: .whitespacesAndNewlines))
}

private func parseInt(_ value: String) -> Int? {
    Int(value.trimmingCharacters(in: .whitespacesAndNewlines))
}

private func parseRPE(_ text: String) -> Double? {
    guard
        let match = rpeValueRegex.firstMatch(in: text, options: [], range: fullRange(of: text)),
        let range = Range(match.range(at: 1), in: text),
        let value = Double(text[range]),
        value >= 1.0,
        value <= 10.0
    else {
        return nil
    }

    return value
}

private func matchesAny(_ patterns: [NSRegularExpression], text: String) -> Bool {
    patterns.contains { $0.firstMatch(in: text, options: [], range: fullRange(of: text)) != nil }
}

private func classifySignal(logText: String, rpe: Double?) -> (String, String) {
    if matchesAny(holdLockPatterns, text: logText) {
        return ("hold_lock", "explicit_keep_instruction")
    }

    if matchesAny(strugglePatterns, text: logText) {
        return ("hold_lock", "struggle_signal")
    }

    if let rpe, rpe >= 9.0 {
        return ("hold_lock", "high_rpe")
    }

    if matchesAny(exceededPatterns, text: logText) {
        return ("progress", "exceeded_signal")
    }

    if let rpe, rpe <= 7.0 {
        return ("progress", "low_rpe")
    }

    return ("neutral", "no_strong_signal")
}

private func normalizeDayName(_ value: String) -> String {
    value.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
}

public func buildProgressionDirectives(priorSupplemental: [String: [PriorSupplementalExercise]]) -> [ProgressionRuleDirective] {
    if priorSupplemental.isEmpty {
        return []
    }

    var directives: [ProgressionRuleDirective] = []

    for dayName in progressionDayOrder {
        for exercise in priorSupplemental[dayName] ?? [] {
            let exerciseName = exercise.exercise.trimmingCharacters(in: .whitespacesAndNewlines)
            if exerciseName.isEmpty {
                continue
            }

            let logText = exercise.log.trimmingCharacters(in: .whitespacesAndNewlines)
            if logText.isEmpty {
                continue
            }

            let parsedRPE = parseRPE(logText)
            let (signal, reason) = classifySignal(logText: logText, rpe: parsedRPE)
            let targetReps = parseInt(exercise.reps)
            let targetLoad = parseFloat(exercise.load)

            directives.append(
                ProgressionRuleDirective(
                    dayName: normalizeDayName(dayName),
                    exerciseName: exerciseName,
                    normalizedExercise: getNormalizer().canonicalKey(exerciseName),
                    signal: signal,
                    reason: reason,
                    holdLock: signal == "hold_lock",
                    targetReps: targetReps,
                    targetLoad: targetLoad,
                    parsedRPE: parsedRPE,
                    sourceLog: logText
                )
            )
        }
    }

    return directives
}

private func formatLoad(_ value: Double?) -> String {
    guard let value else {
        return ""
    }
    if abs(value - round(value)) < 1e-9 {
        return String(Int(round(value)))
    }
    return String(format: "%.3f", value)
        .replacingOccurrences(of: "0+$", with: "", options: .regularExpression)
        .replacingOccurrences(of: "\\.$", with: "", options: .regularExpression)
}

public func formatDirectivesForPrompt(_ directives: [ProgressionRuleDirective], maxLines: Int = 18) -> String {
    if directives.isEmpty {
        return ""
    }

    var lines = ["PROGRESSION DIRECTIVES (DETERMINISTIC FROM PRIOR LOGS):"]
    var holdLines: [String] = []
    var progressLines: [String] = []

    for directive in directives {
        let day = directive.dayName.capitalized
        let logPreview = String(directive.sourceLog.prefix(90))
        if directive.holdLock {
            let reps = directive.targetReps.map(String.init) ?? ""
            let load = formatLoad(directive.targetLoad)
            holdLines.append("- LOCK \(day) | \(directive.exerciseName) -> keep \(reps) reps @ \(load) kg | log: \(logPreview)")
        } else if directive.signal == "progress" {
            progressLines.append("- PROGRESS \(day) | \(directive.exerciseName) -> progression allowed (single-variable change) | log: \(logPreview)")
        }
    }

    lines.append(contentsOf: (holdLines + progressLines).prefix(maxLines))
    return lines.joined(separator: "\n")
}

private func findBestDirective(dayName: String?, exerciseName: String, directives: [ProgressionDirective]) -> ProgressionDirective? {
    guard let dayName else {
        return nil
    }

    let normalizer = getNormalizer()
    let targetDay = normalizeDayName(dayName)

    if let exact = directives.first(where: {
        normalizeDayName($0.dayName ?? "") == targetDay &&
            normalizer.canonicalKey($0.exerciseName) == normalizer.canonicalKey(exerciseName)
    }) {
        return exact
    }

    for directive in directives where normalizeDayName(directive.dayName ?? "") == targetDay {
        if normalizer.areSameExercise(exerciseName, directive.exerciseName) {
            return directive
        }
    }

    return nil
}

public func applyLockedDirectivesToPlan(planText: String, directives: [ProgressionDirective]) -> (String, Int) {
    if planText.isEmpty || directives.isEmpty {
        return (planText, 0)
    }

    let locked = directives.filter { $0.holdLock }
    if locked.isEmpty {
        return (planText, 0)
    }

    let dayRegex = makeRegex("^\\s*##\\s+([A-Z]+DAY)\\b", options: [.caseInsensitive])
    let exerciseRegex = makeRegex("^\\s*###\\s+[A-Z]\\d+\\.\\s*(.+)$", options: [.caseInsensitive])
    let prescriptionRegex = makeRegex("^(\\s*-\\s*)(\\d+)\\s*x\\s*([\\d:]+)\\s*@\\s*([\\d]+(?:\\.\\d+)?)\\s*kg(\\b.*)$", options: [.caseInsensitive])

    var currentDay: String?
    var currentExercise: String?
    var currentDirective: ProgressionDirective?
    var applied = 0
    var lines = planText.components(separatedBy: .newlines)

    for index in lines.indices {
        let line = lines[index]

        if let dayMatch = dayRegex.firstMatch(in: line, options: [], range: fullRange(of: line)),
           let dayRange = Range(dayMatch.range(at: 1), in: line) {
            currentDay = String(line[dayRange]).capitalized
            currentExercise = nil
            currentDirective = nil
            continue
        }

        if let exerciseMatch = exerciseRegex.firstMatch(in: line, options: [], range: fullRange(of: line)),
           let exerciseRange = Range(exerciseMatch.range(at: 1), in: line) {
            currentExercise = String(line[exerciseRange]).trimmingCharacters(in: .whitespacesAndNewlines)
            currentDirective = findBestDirective(dayName: currentDay, exerciseName: currentExercise ?? "", directives: locked)
            continue
        }

        guard let directive = currentDirective else {
            continue
        }

        guard let prescriptionMatch = prescriptionRegex.firstMatch(in: line, options: [], range: fullRange(of: line)),
              let prefixRange = Range(prescriptionMatch.range(at: 1), in: line),
              let setsRange = Range(prescriptionMatch.range(at: 2), in: line),
              let suffixRange = Range(prescriptionMatch.range(at: 5), in: line),
              let targetReps = directive.targetReps,
              let targetLoad = directive.targetLoad
        else {
            continue
        }

        let prefix = String(line[prefixRange])
        let sets = String(line[setsRange])
        let suffix = String(line[suffixRange])
        let newLine = "\(prefix)\(sets) x \(targetReps) @ \(formatLoad(targetLoad)) kg\(suffix)"

        if newLine != line {
            lines[index] = newLine
            applied += 1
        }

        currentDirective = nil
    }

    return (lines.joined(separator: "\n"), applied)
}
